File: pySpec/SpecAnalyze/WavePacketAnalysis/AnalyzeWavePacket.m
classdef AnalyzeWavePacket < handle
   %
   % This file is part of pySpec
   % Copyright (C) 2024  Luis Domenianni
   %
   % This program is free software: you can redistribute it and/or modify
   % it under the terms of the GNU General Public License as published by
   % the Free Software Foundation, either version 3 of the License, or
   % (at your option) any later version.
   %
   % pySpec is distributed in the hope that it will be useful,
   % but WITHOUT ANY WARRANTY; without even the implied warranty of
   % MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   % GNU General Public License for more details.
   %
   % You should have received a copy of the GNU General Public License
   % along with this program.  If not, see <https://www.gnu.org/licenses/>.
   %
    
   properties
       wavelength
       delay
       DeltaOD
       order
       startDelayIndex
       results
   end

   methods
       function results = getResults(this, idx)
           results = this.results{idx};
       end

       function HSVD(this, idx, order)

           % Adapted from Barkhuijsen et al. JMR, vol. 73, 553-557, 1987
           % Delay step, all increments must be even
           delayStep = this.delay(end)-this.delay(end-1);
           
           % Delay of first point
           delay0 = this.delay(this.startDelayIndex);
           
           % Get data
           data = this.DeltaOD(idx, this.startDelayIndex:end);
           
           % generate the Hankel matrix of LxN
           % where L + N = length(data)
           % and 0.5 =< L/N (=1.25) =< 2.0 for optimal results.
                    
           L        = round(1.25/2.25*length(data));
           colFirst = data(1:length(data)-L+1);
           rowLast  = data(length(data)-L+1:length(data));
           H        = hankel(colFirst,rowLast);

           % Singular Value decomposition of the Hankel matrix
           [U,~,~] = svd(H);
           
           % take the truncated matrix and calculate the Z matrix 
           % 'Matrix Computations' by Golub and van Loan, page 3,
           % for the Sherman-Morrison inversion formula.

           u = U(length(data)-L+1,1:order)';
           Z = eye(order)+u*u'/(1-u'*u);
           Z = Z*U(1:length(data)-L,1:order)'*U(2:length(data)-L+1,1:order);
                      
           % calculate signal poles by eigen-decomposition of Z

           poles = eig(Z);
           clear Z

           % calculate the frequencies and time-constants
           frequency=angle(poles)/(2*pi*delayStep);
           ratecon=-log(abs(poles))/delayStep;
           delayConstant=1 ./ratecon;
	
           % now set up the system of equations to calculate the
           % amplitudes and phases...
           % first the matrix
           Z=poles*poles';
           Z=(Z.^(length(data))-1)./(Z-1);

           % now the vector to speed up the routine we use the matlab
           % function 'cumprod' to create a matrix of
           % ascending powers of the conjugate of the
           % vector 'poles'
           vmat=ones(order,length(data));
           vmat(:,2)=conj(poles);

           X=cumprod(vmat');
           X1=cumprod(X);
           vmat =X1';
           clear X X1
           
           for j=1:order
                X=vmat(j,:);     
                rvect(j)=X*data';                                       
                clear X
           end

           % solve the linear system
           svect=(rvect/Z)';

           % calculate the amplitudes and phases
           amplitude = abs(svect).*exp(delay0./delayConstant);
           phase = rem(-(angle(svect)+2*pi*frequency*delay0)*360/(2*pi),360);

           results = [amplitude frequency delayConstant phase];
           [~,I] = sort(abs(results(:,2)));
           results = results(I,:);
           this.results = insert(this.results, idx, {results});
       end
       
       function [delayFit, dataFit] = fitData(this, idx)
           parameter = this.results{idx};

           delayFit = this.delay(this.startDelayIndex:end);
           dataFit = zeros(length(delayFit),1);

           for j = 1:size( parameter, 1 );
               dataFit = dataFit + ( parameter(j,1) * cos( 2 * pi * parameter(j,2) * delayFit +                      ...
                                                           parameter(j,4)*(2*pi)/360)                                ...
                                  .* exp( -delayFit / parameter(j,3) ))'                                               ;
           end
       end
       
       function lowPass(this, idx, freqCutOff)
           indices = find(abs(this.results{idx}(:,2)) > freqCutOff);
           this.results{idx}(indices,:) = [];
       end

       function highPass(this, idx, freqCutOff)
           indices = find(abs(this.results{idx}(:,2)) < freqCutOff);
           this.results{idx}(indices,:) = [];
       end
       end
   
   % Constructor
   methods
       function this = AnalyzeWavePacket(wavelength,...
                                         delay,...
                                         DeltaOD,...
                                         startDelayIndex)
                                    
       this.wavelength          = double(wavelength);
       this.delay               = double(delay);
       this.DeltaOD             = double(DeltaOD);
       this.startDelayIndex     = startDelayIndex;
       this.results             = dictionary();
       
       if isrow(this.wavelength)
           this.wavelength = this.wavelength';
       end
       if iscolumn(this.delay)
           this.delay = this.delay';
       end
       
       if ~isequal(size(this.DeltaOD), [length(this.wavelength) length(this.delay)])
           if isequal(size(this.DeltaOD), [length(this.delay) length(this.wavelength)])
               this.DeltaOD=this.DeltaOD';
           else
               warning('something wrong with DOD matrix')
           end
       end
       end
   end
end

File: pySpec/SpecAnalyze/WavePacketAnalysis/analyzeWavePacket.py
"""
This file is part of pySpec
    Copyright (C) 2024  Markus Bauer

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    pySpec is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

import matlab.engine
import numpy as np

from pySpec.SpecCore.SpecCoreSpectrum.coreTransientSpectrum import TransientSpectrum
from pySpec.SpecCore.SpecCoreSpectrum.coreTransient import Transient


class AnalyzeWavePacket:

    MODULE_PATH = r"./"

    def __init__(self, data: TransientSpectrum or Transient):
        self._data = data
        self._wpa = None
        self._t0 = None
        self._parameter = {}

        self.__eng = matlab.engine.connect_matlab()
        self.__eng.addpath(self.MODULE_PATH, nargout=0)

    @property
    def parameter(self):
        return self._parameter

    @property
    def fit(self):
        y = []
        for i, x in enumerate(self._data.x):
            y.append(self._calculate_fit(self._data.t, i))
            print(f"Fitted Position {x:.1f}, Index {i}!")

        return TransientSpectrum(self._data.x, self._data.t, np.array(y),
                                 self._data.x.unit, self._data.t.unit, self._data.y.unit)
    @property
    def fit_ml(self):
        y = []
        for i, x in enumerate(self._data.x):
            t, y_step = self.__eng.fitData(self._wpa, i+1, nargout=2)
            y.append(np.array(y_step))
            print(f"Fitted Position {x:.1f}, Index {i}!")

        return TransientSpectrum(self._data.x, np.array(t)[0], np.array(y)[..., 0],
                                 self._data.x.unit, self._data.t.unit, self._data.y.unit)

    @property
    def t0(self):
        return self._t0

    def _calculate_fit(self, t, idx):
        parameter = self._parameter[idx]

        print(parameter)

        y = np.zeros_like(t)
        for amp, freq, damp, phase in zip(parameter['amplitude'],
                                          parameter['frequency'],
                                          parameter['delay'],
                                          parameter['phase']):
            y = y + amp * np.cos(2 * np.pi * freq * t + phase * (2*np.pi/360)) * np.exp(-t/damp)

        return y

    def cleanup(self):
        self.__eng.quit()

    def hsvd(self, order: int = 25, t0: float or None = None, t0_idx: int = 0):
        if t0 is not None:
            t0_idx, _ = self._data.t.closest_to(t0)

        self._t0 = (self._data.t[t0_idx], t0_idx)

        self._wpa = self.__eng.AnalyzeWavePacket(self._data.x.array,
                                                 self._data.t.array,
                                                 self._data.y.array,
                                                 t0_idx + 1)

        for i, x in enumerate(self._data.x):
            self.__eng.HSVD(self._wpa, i+1, order, nargout=0)
            print(f"Analyzed Position {x}, Index {i}!")
        self._retrieve_parameter()

    def _retrieve_parameter(self):
        for i, _ in enumerate(self._data.x):
            result = np.array(self.__eng.getResults(self._wpa, i+1, nargout=1))

            self._parameter[i] = {'amplitude': result[:, 0],
                                  'frequency': result[:, 1],
                                  'delay': result[:, 2],
                                  'phase': result[:, 3]}

    def low_pass(self, frequency):
        if self._wpa is None:
            raise ValueError()

        for i, _ in enumerate(self._data.x):
            mask = abs(self._parameter[i]['frequency']) < frequency
            for key, val in self._parameter[i].items():
                self._parameter[i][key] = val[mask]

    def high_pass(self, frequency):
        if self._wpa is None:
            raise ValueError()

        for i, _ in enumerate(self._data.x):
            mask = abs(self._parameter[i]['frequency']) > frequency
            for key, val in self._parameter[i].items():
                self._parameter[i][key] = val[mask]

    def isolate(self, frequency, margin=0.05):
        self._retrieve_parameter()

        freq_lo = (1 - margin) * frequency
        freq_hi = (1 + margin) * frequency

        self.high_pass(freq_lo)
        self.low_pass(freq_hi)


if __name__ == '__main__':
    from pySpec.SpecImport.importTimeResolvedFTIR import ImportTimeResolvedFTIR
    import matplotlib.pyplot as plt

    paths = ['']

    f = ImportTimeResolvedFTIR.from_files(paths, 190)

    g = f.average

    hsvd = AnalyzeWavePacket(g)
    hsvd.hsvd(25, 50)
    hsvd.isolate(1 / 500, 0.5)

    fit_res = hsvd.fit
    clean_data = g - fit_res

    fig, ax = plt.subplots(1, 1)
    for i in range(0, len(g.x), 20):
        ax.plot(g.t, g.transient[i].y + i/2000, marker='o', color='grey', ls='')
        ax.plot(fit_res.t, fit_res.transient[i].y + i/2000, lw=2)
        ax.plot(clean_data.t, clean_data.transient[i].y + i/2000)

    plt.show()
